### PR TITLE
Description feature

### PR DESCRIPTION
**Pull request by grdw**
This PR adds a feature to the SAX XMLParser to parse out the first p-tag out of a page and concats all child-tags' `text` attributes to the first open p-tag. The XMLParser now pauses when a p-tag and img-tag are both found and closed.

**To summarize the changes:**
*on El:*
The `El` struct has been turned into a class and received an additional `open` boolean. This boolean determines if a HTML element is open (f.e. `<p>...`) or closed (f.e. `</p> ...`). The SAX parser knows when an element opens and closes by listening to the `start_element(..)` and `end_element(..)` callbacks. An example of this behavior can be seen [in this spec](https://github.com/WeTransfer/richurls/blob/df8be314e52124bbb59c1e683438fa62d8251413/spec/lib/xml_handler_spec.rb#L50-L80).

*on stopping the XMLParser:*
The way the SAX XMLParser stops is now done very differently. First it was done by 'finding' the first `` tag in the page. If the parser found it, it knew it could stop. Now we have a p-tag which makes the stopping a lot more complex. The reason for this is because you can have 3 situations:

The p-tag is before the img-tag:

```html
<p></p>
<img/>
```

or

The img-tag is before the p-tag:

```html
<img/>
<p></p>
```

or

The img-tag is inside of the p-tag:

```html
<p>
  <img/>
</p>
```

The third situation is the entire reason we are looking to see if a tag is open/closed. Both need to be 'closed' in order for the SAX parser to stop.

*on stopping the XMLParser (performance):*
Performance wise I added a method to make sure we only find one of either a p-tag or img-tag which why there's a 10.008 line fixture in this PR. 

**About the benchmarking:**

*For a Youtube page:*

```
ox html parsing (old):      431.5 i/s
     ox html parsing:       22.1 i/s - 19.54x  slower
```

Benchmarking this to the old implementation for a Youtube video page it's 19.54x slower. This is quite a penalty. The reason for this is purely the way we determine how and when to stop. However considering that the previous speed increase on the old XMLParser was pretty much non-visible in the charts, I think it's a penalty we can manage. The difference between the older XMLparser (non breaking) and this one is:

```
ox html parsing (older):    25.1 i/s
     ox html parsing:       22.7 i/s - 1.11x  slower
```

1.11x slower is understandable considering the added complexity.

*For WeTransfer.com:*

```
ox html parsing (old):     1345.0 i/s
     ox html parsing:      762.9 i/s - 1.76x  slower
```

1.76x slower is acceptable seeing how fast a single iteration is finished.

---

Fixes: #28

### CODE REVIEW

File: lib/body_decorator.rb [hunk 1]
```diff
@@ -36,8 +42,8 @@ def initialize(url, body)
     end
 
     def decorate
-      PARSERS.each_with_object({}) do |(key, parser), object|
-        object[key] = parser.call(@xml, @url)&.force_encoding('UTF-8')
+      PARSERS.transform_values do |parser|
```
Comment (lorenzograndi4) [inline lib/body_decorator.rb]: 👏

File: lib/el.rb [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+      @attributes = {}
+    end
+
+    def add(key, value)
+      return if @attributes[key]
+
+      @attributes[key] = value
+    end
+
+    def append_text(str)
```
Comment (lorenzograndi4) [inline lib/el.rb]: I think I got lost. When/why do you append stuff like this?

File: lib/xml_handler.rb [hunk 3]
```diff
@@ -17,48 +18,67 @@ class XMLHandler < ::Ox::Sax
...
     def initialize
       @elements = []
+      @counts = Set.new
+      @breaks = []
     end
 
-    def find(name, attrs = {})
-      @elements.detect do |el|
+    def find(tag, attrs = {})
+      elements.detect do |el|
```
Comment (lorenzograndi4) [inline lib/xml_handler.rb]: I find it confusing that you're using `elements` here and `@elements` everywhere else. Is there a reason for that?

Comment (grdw) [reply lib/xml_handler.rb]: No, good spot!

File: lib/el.rb (hunk 2 above)
Comment (grdw) [reply lib/el.rb]: Whenever the `XMLHandler` comes accross a piece of `text()`, so [at this line](https://github.com/WeTransfer/richurls/blob/description-feature/lib/xml_handler.rb#L74).

**Pull request merged**